modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -48,37 +52,51 @@
 
 (defn- ^Iterable list-tables
   "Fetch all tables (new pages are loaded automatically by the API)."
-  (^Iterable [{{:keys [project-id dataset-id]} :details, :as database}]
-    (let [bq (database->client database)]
-      (list-tables bq project-id dataset-id)))
-
-  (^Iterable [^BigQuery client, ^String project-id ^String dataset-id]
-    {:pre [client (not (str/blank? dataset-id))]}
-    (.iterateAll (.listTables client (DatasetId/of project-id dataset-id) (u/varargs BigQuery$TableListOption)))))
+  (^Iterable [{{:keys [project-id dataset-filters-type dataset-filters-patterns]} :details, :as database}]
+    (list-tables (database->client database)
+                 (or project-id (bigquery.common/database-details->credential-project-id (:details database)))
+                 dataset-filters-type
+                 dataset-filters-patterns))
+
+  (^Iterable [^BigQuery client ^String project-id ^String filter-type ^String filter-patterns]
+    (let [datasets (.listDatasets client project-id (u/varargs BigQuery$DatasetListOption))
+          inclusion-patterns (when (= "inclusion" filter-type) filter-patterns)
+          exclusion-patterns (when (= "exclusion" filter-type) filter-patterns)]
+      (apply concat (for [^Dataset dataset (.iterateAll datasets)
+                          :let [^DatasetId dataset-id (.. dataset getDatasetId)]
+                          :when (describe-database/include-schema? (.getDataset dataset-id)
```

Comment:
why is there a call to `(.getDataset dataset-id)` if we have a `dataset` and the id comes from `(.. dataset getDatasetId)`?

Reply:
`dataset-id` is actually an object (of type `DatasetId`), and `.getDatasetId` is the Java method to return the String value of the ID.  The reason I capture it separately is because the `DatasetId` object also implicitly has a `project-id` associated to it, so the `listTables` operation below can accept that object and use the embedded project-id/dataset-id tuple.